### PR TITLE
feat(framework) Add Ffs factory

### PR DESCRIPTION
**Pull request by charlesbvll**
### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.ai/docs/writing-documentation.html)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### CODE REVIEW

File: src/py/flwr/server/superlink/ffs/ffs_factory.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+from .ffs import Ffs
+
+
+class FfsFactory:
+    """Factory class that creates Ffs instances.
+
+    Parameters
+    ----------
+    base_dir : str
+        The base directory to store the objects.
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/ffs/ffs_factory.py:33]:
```suggestion
        The base directory used by DiskFfs to store objects.
```

File: src/py/flwr/server/superlink/ffs/ffs_factory.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        The base directory used by DiskFfs to store objects.
+    """
+
+    def __init__(self, base_dir: str) -> None:
+        self.base_dir = base_dir
+        self.ffs_instance: Optional[Ffs] = None
+
+    def ffs(self) -> Ffs:
+        """Return a Ffs instance and create it, if necessary."""
+        # SqliteState
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/ffs/ffs_factory.py:42]:
```suggestion
```

File: src/py/flwr/server/superlink/ffs/ffs_factory.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    """
+
+    def __init__(self, base_dir: str) -> None:
+        self.base_dir = base_dir
+        self.ffs_instance: Optional[Ffs] = None
+
+    def ffs(self) -> Ffs:
+        """Return a Ffs instance and create it, if necessary."""
+        ffs = DiskFfs(self.base_dir)
+        log(DEBUG, "Using Disk Flower File System")
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/ffs/ffs_factory.py:43]:
```suggestion
        log(DEBUG, "Using DiskFfs")
```

File: src/py/flwr/server/superlink/ffs/ffs_factory.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    Parameters
+    ----------
+    base_dir : str
+        The base directory used by DiskFfs to store objects.
+    """
+
+    def __init__(self, base_dir: str) -> None:
+        self.base_dir = base_dir
+        self.ffs_instance: Optional[Ffs] = None
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/ffs/ffs_factory.py:38]: This is not used

Comment (charlesbvll) [reply src/py/flwr/server/superlink/ffs/ffs_factory.py:38]: @danieljanes Oh right, I updated it

File: src/py/flwr/server/superlink/ffs/ffs_factory.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    def ffs(self) -> Ffs:
+        """Return a Ffs instance and create it, if necessary."""
+        if self.ffs_instance:
+            return self.ffs_instance
+
+        self.ffs_instance = DiskFfs(self.base_dir)
+        log(DEBUG, "Using DiskFfs")
+        return self.ffs_instance
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/ffs/ffs_factory.py:47]:
```suggestion
        """Return a Ffs instance and create it, if necessary."""
        if not self.ffs_instance:
            log(DEBUG, "Initialize DiskFfs")
            self.ffs_instance = DiskFfs(self.base_dir)

        log(DEBUG, "Using DiskFfs")
        return self.ffs_instance
```

File: src/py/flwr/server/superlink/ffs/ffs_factory.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    """
+
+    def __init__(self, base_dir: str) -> None:
+        self.base_dir = base_dir
+        self.ffs_instance: Optional[Ffs] = None
+
+    def ffs(self) -> Ffs:
+        """Return a Ffs instance and create it, if necessary."""
+        if not self.ffs_instance:
+            log(DEBUG, "Initialize DiskFfs")
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/ffs/ffs_factory.py:43]:
```suggestion
            log(DEBUG, "Initializing DiskFfs")
```

File: src/py/flwr/server/superlink/ffs/ffs_factory_test.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+# limitations under the License.
+# ==============================================================================
+"""Test Ffs factory."""
+
+from .ffs import Ffs
+from .ffs_factory import FfsFactory
+
+
+def test_disk_ffs_factory() -> None:
+    """Test add_loss_distributed."""
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/ffs/ffs_factory_test.py:22]: c&p?

**Review by danieljanes — APPROVED** [review]

**Pull request merged**